packages/from-schema/src/index.ts

```diff
@@ -158,68 +158,92 @@ const run: (schemaInput: GraphQLSchema, optionsInput: IInternalOptions) => strin
     );
   };
 
+  type ResolvedInterfaceValue = {
+    value: ResolvedInterfaceValue | string;
+    description?: string;
+    deprecation?: string;
+    isList: boolean;
+    isNonNull: boolean;
+  };
+
+  type ResolveInterfaceName = (type: GraphQLInputType | GraphQLType, isNonNull: boolean) => ResolvedInterfaceValue;
+
   /**
    * TODO
    * - add support for custom types (via optional json file or something)
-   * - allow this to return metadata for Non Null types
    */
-  const resolveInterfaceName: (type: GraphQLInputType | GraphQLType) => string = type => {
+  const resolveInterfaceName: ResolveInterfaceName = (type, isNonNull = false) => {
     if (isList(type)) {
-      return wrapList(resolveInterfaceName((type).ofType));
-    } else if (isNonNullable(type)) {
-      return `!${resolveInterfaceName((type).ofType)}`;
-    } else if (isScalar(type)) {
-      return TYPE_MAP[type.name] || TYPE_MAP.__DEFAULT;
-    } else if (isAbstractType(type)) {
-      return generateTypeName(type.name);
-    } else if (isEnum(type)) {
-      return generateEnumName(type.name);
-    } else {
-      return generateInterfaceName(type.name);
+      return {
+        value: resolveInterfaceName(type.ofType, false),
+        isList: true,
+        isNonNull
+      };
+    }
+    if (isNonNullable(type)) {
+      return resolveInterfaceName(type.ofType, true);
+    }
+    if (isScalar(type)) {
+      return {
+        value: DEFAULT_TYPE_MAP[type.name] || DEFAULT_TYPE_MAP.__DEFAULT,
+        isList: false,
+        isNonNull
+      };
+    }
+    if (isAbstractType(type)) {
+      return {
+        value: generateTypeName(type.name),
+        isList: false,
+        isNonNull
+      };
+    }
+    if (isEnum(type)) {
+      return {
+        value: generateEnumName(type.name),
+        isList: false,
+        isNonNull
+      };
     }
+    return {
+      value: generateInterfaceName(type.name),
+      isList: false,
+      isNonNull
+    };
   };
 
-  interface IInterfaceMetadata {
-    name: string;
-    showNullabilityAttribute: boolean;
-    isNonNull: boolean;
-  }
+  type TypePrinter = (val: ResolvedInterfaceValue | string, supportsNullability: boolean) => string;
```

Comment:
when would `val` be a `string`?

Reply:
The "final" value of the `ResolvedInterfaceValue.value` is always a string with its appropriate type (scalar, abstract, enum, plain, ...) we need to be able to simply print that.

The `typePrinter` function is recursively going deeper in the type structure, and passing in its `val.value` so eventually we'll end up with just a plain string.

So basically our type value structure might look something like:

```jsonc
{
  value: { // <-- the initial type is an object.
    value: 'string', // <-- the inner/final type is a string.
    isList: false,
    isNonNull: true
  },
  isList: true,
  isNonNull: false
}
```

Which is essentially `[String!]` in GraphQL. It's important to note that sometimes it could be even deeper than this, eg. if we have a double-array such as `[[Int!]]` or similar.

Reply:
If you can think of a better way to implement this I'm open for suggestions, however! This was just what made most sense for me on top of my head. :)

Reply:
Ah makes sense 👍